src/Learners/Core/IndexedSet.cs

```diff
@@ -66,7 +66,8 @@ public IndexedSet(IEnumerable<T> elements) : this()
         /// from a reader of a binary stream.
         /// </summary>
         /// <param name="reader">The reader to load the indexed set from.</param>
-        public IndexedSet(IReader reader) : this()
+        /// <param name="parseEntity">Optional custom parsing for items in the set.</param>
+        public IndexedSet(IReader reader, Func<string, T> parseEntity = null) : this()
```

Comment:
In general only the caller knows the entry type; so this means having some sort of ReaderWrapper type that adds specific serialization logic for typeof(T) to one of the generic IReader implementations? Or are you suggesting having one of the generic implementations handle all types? This isn't feasible without writing our own version of something like BinaryFormatter. I suppose a third option would be to go through each usage of IndexedSet and add extra logic into an IReader implementation so that it handles all of the entry types. In these options a third party would need to implement their own IReader.

Another option is to change the semantics of this type: currently T can be anything and that's the issue because "serialize and deserialize ANYTHING" is problematic. But we could restrict the types that are fully supported to one of the types IReader can take explicitly (such as int, string, etc.), and throw an exception otherwise.

Reply:
The first thing you said.